core/message_parsers.py

```diff
@@ -0,0 +1,350 @@
+from dataclasses import dataclass
+from typing import List, Optional
+from datetime import datetime
+import re
+import matplotlib.pyplot as plt
+import io
+import pandas as pd
+import numpy as np
+
+import discord
+
+class HypixelRank:
+    # Using emojis or special characters to represent different ranks
+    RANK_FORMATS = {
+        'VIP': '🟢',      # Green circle for VIP
+        'VIP+': '🟢⭐',    # Green circle with star for VIP+
+        'MVP': '🔷',      # Blue diamond for MVP
+        'MVP+': '🔷⭐',    # Blue diamond with star for MVP+
+        'MVP++': '🟡⭐',   # Gold circle with star for MVP++
+        'ADMIN': '🔴',    # Red circle for ADMIN
+        'HELPER': '💙',   # Blue heart for HELPER
+        'MODERATOR': '💚' # Green heart for MODERATOR
+    }
+
+    @staticmethod
+    def format_rank(rank: str) -> str:
+        rank = rank.upper() if rank else ''
+        emoji = HypixelRank.RANK_FORMATS.get(rank, '')
+        return f'{emoji}**[{rank}]**' if rank else ''
+
+@dataclass
+class GuildMember:
+    name: str
+    rank: Optional[str] = None
+    experience: Optional[int] = None
+
+@dataclass
+class GuildRole:
+    name: str
+    members: List[GuildMember]
+
+@dataclass
+class TopEntry:
+    member: GuildMember
+    experience: int
+    position: int
+
+class GuildMessageParser:
+    def __init__(self, message: str):
+        self.raw_message = message
+        self.guild_name = ""
+        self.total_members = 0
+        self.online_members = 0
+        self.offline_members = 0
+        self.roles = []
+        self.top_entries = []
+        self.date = None
+        
+    def parse(self) -> str:
+        # Determine message type and parse accordingly
+        if "Top" in self.raw_message:
+            print("Command: Top")
+            return self._parse_top_message()
+        elif "Total Members:" in self.raw_message:
+            if "Offline Members:" in self.raw_message:
+                print("Command: Online")
+                return self._parse_online_message()
+            else:
+                print("Command: List")
+                return self._parse_list_message()
+        elif "MOTD" in self.raw_message:
+            print("Command: Info")
+            return self._create_guild_stats_embed()
+        else:
+            return "NaN"
+            
+    def _clean_rank(self, rank: str) -> str:
+        rank = rank.strip('[]').strip()
```

Comment:
Shouldn't this be `.lstrip("[")`?

Reply:
Alternatively, use capture groups in regex instead

Reply:
Ok, will do this in the next commit